### ISSUE TITLE
[AOT Autograd] operator decomposition does not respect training=False flag for torch.nn.functional.dropout

### ISSUE THREAD

**Issue by youkaichao** [pytorch/pytorch#115249]
### 🐛 Describe the bug

When specifying `training=False`, `torch.nn.functional.dropout` is essentially a no-operation. However, after AOTAutograd, `torch.nn.functional.dropout` always clones the input.

Here is the code:

```python
import torch
def f(x):
    y = torch.nn.functional.dropout(x, p=0.5, training=False)
    return y

x = torch.randn(5)
y = f(x)
assert y is x

f_eager = torch.compile(f, backend="eager")
y_eager = f_eager(x)

assert y_eager is x

f_aot_eager = torch.compile(f, backend="aot_eager")
y_aot_eager = f_aot_eager(x)

assert y_aot_eager is not x
```

As we can see, the eager mode, compile with eager mode, both treat `torch.nn.functional.dropout` as a no-operation, and returns the **same** tensor input as output. However, after compile with aot autograd, the output is another tensor.

Inspecting the logging output, I find that aot autograd uses `torch.ops.aten.clone.default`. The code responsible for this behavior is:

https://github.com/pytorch/pytorch/blob/fcf6a76108be8e7b6db528b631a7b8ebdc7470ac/torch/_decomp/decompositions.py#L1089

Is it desired behavior or it is a bug that should be fixed?

### Versions

Nightly build version `2.2.0.dev20231127`.

**Comment by youkaichao** [comment]
Might be relevant: @Chillee

**Comment by youkaichao** [comment]
Is this a desired behavior? Simply disabling the `clone` breaks so many tests at https://github.com/pytorch/pytorch/pull/115258 .

**Comment by youkaichao** [comment]
Fixing this bug requires modification from inductor side cc @jansel 

If we want to fix this bug, some function call without dropout and some function call with dropout in inference mode has the same graph, which fails this assertion:

https://github.com/pytorch/pytorch/blob/1c1f2bbe8ac5e1642aebd2ef7293e972fcbd048e/torch/_inductor/pattern_matcher.py#L978

To avoid the assertion, we might need to record search/replace pairs, and compare replace graph with existing one to allow `register_replacement` to be called with the same pair of search/replace pairs for more than once.

Another solution is to remove the assertion though. But it might cause confusion if one occasionally registers the same search graph with different replace graph.

Would like to hear opinions from @jansel and @Chillee .

**Comment by bdhirsh** [comment]
Related: https://github.com/pytorch/pytorch/issues/97894

I think one issue with this approach is that:

(1) `native_dropout` is a core aten operator, which needs to have clear aliasing and mutation semantics in order to be correct w.r.t. autograd and functionalization

(2) removing the clone() in the inference case will violate this contract: `native_dropout` would return an input directly in the inference case, but a fresh tensor in the training case

I think instead of handling this in the decomp, it's on the compiler to try to optimize the clone away (for example, inductor can see that removing the clone wouldn't affect the aliasing behavior between graph inputs and outputs, and decide that it's safe to remove)

**Comment by bdhirsh** [comment]
yep- the `aot_eager` isn't meant to be fast (it's usually slower than eager mode anyway), but it is meant to be always correct. It's been useful for debugging compiler bugs - if we get wrong results with default `torch.compile` but not with `torch.compile(backend="aot_eager")`, that usually points to an inductor-specific bug.

**Comment by bdhirsh** [comment]
yep, this is still happening today. Inductor handles it by running a graph pass that removes unnecessary clones. I'm not actually sure if this is (still) a problem for export. Maybe @tugsbayasgalan or @angelayi know?

**Comment by aorenste** [comment]
This issue still reproduces on PyTorch 2.12.0a0 (nightly).

(Analysis done by Claude.)
